cpp/CMakeLists.txt

```diff
@@ -693,7 +707,11 @@ if(NOT APPLE AND NOT MSVC)
 endif()
 
 set(ARROW_ALL_SRCS
-  ${ARROW_SRCS})
+  ${ARROW_PARQUET_SRCS})
+
+if(NOT ARROW_BUILD_PARQUET_DEPS_ONLY)
+  list(APPEND ARROW_ALL_SRCS ${ARROW_SRCS})
+endif()
```

Comment:
This may grow a little bit difficult to maintain. Per my comment below about the compression code organization, once there are variables to toggle including LZ4 and ZSTD we can have code like:

```
if (ARROW_WITH_LZ4)
  SET(ARROW_LINK_LIBS lz4_static ${ARROW_LINK_LIBS)
  SET(ARROW_SRCS src/arrow/util/compression_lz4.cc ${ARROW_SRCS})
endif()
```

I don't think there's harm in including the additional object code in `libarrow.a` so long as no additional transitive dependencies are introduced